pkg/apis/auditregistration/types.go

```diff
@@ -0,0 +1,282 @@
+	// to turn up in a new cluster.
+	//
+	// The scheme must be "https"; the URL must begin with "https://".
+	//
+	// A path is optional, and if present may be any string permissible in
+	// a URL. You may use the path to pass an arbitrary string to the
+	// webhook, for example, a cluster identifier.
+	//
+	// Attempting to use a user or basic auth e.g. "user:password@" is not
+	// allowed. Fragments ("#...") and query parameters ("?...") are not
+	// allowed, either.
+	//
+	// +optional
+	URL *string
+
+	// `service` is a reference to the service for this webhook. Either
+	// `service` or `url` must be specified.
+	//
+	// If the webhook is running within the cluster, then you should use `service`.
+	//
+	// Port 443 will be used if it is open, otherwise it is an error.
+	//
+	// +optional
+	Service *ServiceReference
+
+	// `caBundle` is a PEM encoded CA bundle which will be used to validate
+	// the webhook's server certificate.
+	// +optional
+	CABundle []byte
+}
+
+// ServiceReference holds a reference to Service.legacy.k8s.io
+type ServiceReference struct {
+	// `namespace` is the namespace of the service.
+	// Required
+	Namespace string
+
+	// `name` is the name of the service.
+	// Required
+	Name string
+
+	// `path` is an optional URL path which will be sent in any request to
+	// this service.
+	// +optional
+	Path *string
+}
+
+// Policy defines the configuration of audit logging, and the rules for how different request
+// categories are logged.
+type Policy struct {
+	// Rules specify the audit Level a request should be recorded at.
+	// A request may match multiple rules, in which case the FIRST matching rule is used.
+	// The default audit level is None, but can be overridden by a catch-all rule at the end of the list.
+	// PolicyRules are strictly ordered.
+	Rules []PolicyRule
+
+	// OmitStages is a list of stages for which no events are created. Note that this can also
+	// be specified per rule in which case the union of both are omitted.
+	// +optional
+	OmitStages []Stage
+}
+
+// PolicyRule maps requests based off metadata to an audit Level.
+// Requests must match the rules of every field (an intersection of rules).
+type PolicyRule struct {
+	// The Level that requests matching this rule are recorded at.
+	Level Level
+
+	// The users (by authenticated user name) this rule applies to.
+	// An empty list implies every user.
+	// +optional
+	Users []string
+	// The user groups this rule applies to. A user is considered matching
+	// if it is a member of any of the UserGroups.
+	// An empty list implies every user group.
+	// +optional
+	UserGroups []string
+
+	// The verbs that match this rule.
+	// An empty list implies every verb.
+	// +optional
+	Verbs []string
+
+	// Rules can apply to API resources (such as "pods" or "secrets"),
+	// non-resource URL paths (such as "/api"), or neither, but not both.
+	// If neither is specified, the rule is treated as a default for all URLs.
```

Comment:
This doesn't make much sense, it contradicts the preceding sentence.

Reply:
I had to read twice but it is valid 😉

Reply:
Maybe this would be more clear:
> A PolicyRule can either specify selectors for API resources (such as "pods" or "secrets")
> or non-resource URL paths (such as "/api"), but not both.
> If neither is specified, the rule applies to all resource & non-resource requests.

(BTW, @lavalamp I wrote the original API, which the policy is a copy of)

Reply:
Apologies for pedantry, but the "but not both" part still seems directly contradicted by the next sentence. I think what is meant is the below?

```
A PolicyRule can either specify selectors for API resources (such as "pods" or "secrets")
or non-resource URL paths (such as "/api").
If neither is specified, the rule applies to all resource & non-resource requests.
However, you may not select specific non-resource paths and resource paths in the same rule
```

I get that it is obscure that people would want to select both in the same rule, but why go out of your way to forbid it?

Reply:
All the other fields interact additively (an incoming request must match one of the specified values for all the filtering fields). Resource and non-resource requests have mutually exclusive attributes, so having a single rule contain filtering fields that don't actually get applied to a ln incoming request is confusing

Reply:
BTW, this was originally copied from RBAC: https://github.com/kubernetes/kubernetes/blob/bd2370d2ae629ef4826827bcb93b3b4cbfb9f406/pkg/apis/rbac/types.go#L60